mllib/src/main/scala/org/apache/spark/ml/evaluation/MultilabelClassificationEvaluator.scala

```diff
@@ -0,0 +1,148 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.evaluation
+
+import org.apache.spark.annotation.{Experimental, Since}
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.util._
+import org.apache.spark.mllib.evaluation.MultilabelMetrics
+import org.apache.spark.sql.Dataset
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+
+/**
+ * :: Experimental ::
+ * Evaluator for multi-label classification, which expects two input
+ * columns: prediction and label.
+ */
+@Since("3.0.0")
+@Experimental
+class MultilabelClassificationEvaluator @Since("3.0.0") (override val uid: String)
+  extends Evaluator with HasPredictionCol with HasLabelCol
+    with DefaultParamsWritable {
+
+  import MultilabelClassificationEvaluator.supportedMetricNames
+
+  @Since("3.0.0")
+  def this() = this(Identifiable.randomUID("mlcEval"))
+
+  /**
+   * param for metric name in evaluation (supports `"f1Measure"` (default), `"subsetAccuracy"`,
+   * `"accuracy"`, `"hammingLoss"`, `"precision"`, `"recall"`, `"precisionByLabel"`,
+   * `"recallByLabel"`, `"f1MeasureByLabel"`, `"microPrecision"`, `"microRecall"`,
+   * `"microF1Measure"`)
+   * @group param
+   */
+  @Since("3.0.0")
+  final val metricName: Param[String] = {
+    val allowedParams = ParamValidators.inArray(supportedMetricNames)
+    new Param(this, "metricName", "metric name in evaluation " +
+      s"${supportedMetricNames.mkString("(", "|", ")")}", allowedParams)
+  }
+
+  /** @group getParam */
+  @Since("3.0.0")
+  def getMetricName: String = $(metricName)
+
+  /** @group setParam */
+  @Since("3.0.0")
+  def setMetricName(value: String): this.type = set(metricName, value)
+
+  setDefault(metricName -> "f1Measure")
+
+  @Since("3.0.0")
+  final val label: DoubleParam = new DoubleParam(this, "label",
+    "The label whose metric will be computed in precisionByLabel|recallByLabel|" +
+      "f1MeasureByLabel. Must be >= 0. The default value is 0.",
+    ParamValidators.gtEq(0.0))
+
+  @Since("3.0.0")
+  def getLabel: Double = $(label)
+
+  @Since("3.0.0")
+  def setLabel(value: Double): this.type = set(label, value)
+
+  setDefault(label -> 0.0)
+
+  /** @group setParam */
+  @Since("3.0.0")
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  /** @group setParam */
+  @Since("3.0.0")
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+
+  @Since("3.0.0")
+  override def evaluate(dataset: Dataset[_]): Double = {
+    val schema = dataset.schema
+    SchemaUtils.checkColumnTypes(schema, $(predictionCol),
+      Seq(ArrayType(DoubleType, false), ArrayType(DoubleType, true)))
```

Comment:
In the Python side, the default datatype of array columns create by `spark.createDataFrame` or `.toDF` is ArrayType(_, true), to pass the python tests, we need to allow `ArrayType(DoubleType, true)` here.